base/dataset_prepare.go

```diff
@@ -12,66 +13,117 @@ import (
-
-	if curr, err = dsfs.LoadDataset(ctx, r.Store(), currPath); err != nil {
-		return
+	// Validate that username is our own, it's not valid to try to save a dataset with someone
+	// else's username. Without this check, base will replace the username with our own regardless,
+	// it's better to have an error to display, rather than silently ignore it.
+	if ref.Username != "" && ref.Username != "me" && ref.Username != pro.Peername {
+		return ref, isNew, fmt.Errorf("cannot save using a different username than %q", pro.Peername)
 	}
-	if curr.BodyPath != "" {
-		var body qfs.File
-		body, err = dsfs.LoadBody(ctx, r.Store(), curr)
-		if err != nil {
-			return
+	ref.Username = pro.Peername
+
+	// attempt to resolve the reference
+	if _, resolveErr := resolver.ResolveRef(ctx, &ref); resolveErr != nil {
+		if !errors.Is(resolveErr, dsref.ErrRefNotFound) {
+			return ref, isNew, resolveErr
+		}
+	} else if resolveErr == nil {
+		if newName {
+			// Name was explicitly given, with the --new flag, but the name is already in use.
+			// This is an error.
+			return ref, isNew, qerr.New(ErrNameTaken, "dataset name has a previous version, cannot make new dataset")
 		}
-		curr.SetBodyFile(body)
+
+		if badCaseErr != nil {
+			// name already exists but is a bad case, log a warning and then continue.
+			log.Error(badCaseErr)
+		}
+
+		// we have a valid previous reference & an initID, return!
+		return ref, isNew, nil
 	}
 
-	// Load a mutable copy of the dataset because most of the save path assuming we are doing
-	// a patch update to the current head, and not a full replacement.
-	if mutable, err = dsfs.LoadDataset(ctx, r.Store(), currPath); err != nil {
-		return
+	// at this point we're attempting to create a new dataset
+	isNew = true
+
+	// If dataset name is using bad-case characters, and is not yet in use, fail with error.
+	if badCaseErr != nil {
+		return ref, isNew, badCaseErr
+	}
+	if !dsref.IsValidName(ref.Name) {
+		return ref, isNew, fmt.Errorf("invalid dataset name: %s", ref.Name)
 	}
 
-	// TODO(dustmop): Stop removing the transform once we move to apply, and untangle the
-	// save command from applying a transform.
-	// remove the Transform & commit
-	// transform & commit must be created from scratch with each new version
-	mutable.Transform = nil
-	mutable.Commit = nil
-	return
+	ref.InitID, err = book.WriteDatasetInit(ctx, ref.Name)
+	return ref, isNew, err
 }
 
-// MaybeInferName infer a name for the dataset if none is set
-func MaybeInferName(ds *dataset.Dataset) string {
-	if ds.Name == "" {
-		filename := ds.BodyFile().FileName()
-		basename := strings.TrimSuffix(filename, filepath.Ext(filename))
-		return dsref.GenerateName(basename, "dataset_")
+// GenerateAvailableName creates a name for the dataset that is not currently in use
+func GenerateAvailableName(ctx context.Context, pro *profile.Profile, resolver dsref.Resolver, prefix string) string {
+	lookup := &dsref.Ref{Username: pro.Peername, Name: prefix}
+	counter := 1
```

Comment:
Before, this used `counter := 0` so that the first attempted name is "dataset_1" instead of "dataset_2". Wondering if this is an intended change. So if I drag a csv file named "stats.csv" into Desktop (for example), I get "me/stats" then doing it again gives me "me/stats_2"? I was thinking it was better to start generating names with "..._1" but maybe that's programmer-think getting in the way.

Regardless, please add a comment stating the assumption (whichever way we go) to make it explicit that it's expected behavior.

Reply:
Yeah, here I'm proposing matching the OS X convention of starting with `_2`. From a UX perspective, the `_1` is implied as the first "stats" dataset in your example. I'll add a comment